File: 백준/Gold/1240. 노드사이의 거리/노드사이의 거리.py
import sys
from collections import deque

input = sys.stdin.readline

nv, np = map(int, input().split())
edges = [[] for _ in range(nv+1)]
for _ in range(nv-1):
    v1, v2, dist = map(int, input().split())
    edges[v1].append((v2, dist))
    edges[v2].append((v1, dist))

def bfs(start, target):
    q = deque()
    visited = [False] * (nv+1)
    q.append((start, 0))
    visited[start] = True
    while q:
        cur, dist = q.popleft()
        for nxt, delta in edges[cur]:
            if nxt == target:
                return delta + dist
            if not visited[nxt]:
                visited[nxt] = True
                q.append((nxt, delta + dist))
    return 0

for _ in range(np):
    v1, v2 = map(int, input().split())
    print(bfs(v1, v2))

File: 백준/Gold/1240. 노드사이의 거리/노드사이의 거리.java
import java.util.*;
import java.io.*;

public class Main {
  public static class Node {
    public int n, cost;
    public Node(int n, int cost) {
      this.n = n;
      this.cost = cost;
    }
  }

  public static int nv;
  public static List<List<Node>> edges = new ArrayList<List<Node>>();

  public static int bfs(int start, int end) {
    Queue<Node> q = new ArrayDeque<>();
    boolean[] visited = new boolean[nv+1];
    q.offer(new Node(start, 0));
    visited[start] = true;
    Node cur;
    while (!q.isEmpty()) {
      cur = q.poll();
      for (Node nxt : edges.get(cur.n)) {
        if (visited[nxt.n]) continue;
        if (nxt.n == end) return cur.cost + nxt.cost;
        visited[nxt.n] = true;
        q.offer(new Node(nxt.n, cur.cost + nxt.cost));
      }
    }
    return 0;
  }

  public static void main(String[] args) throws Exception {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    StringTokenizer st = new StringTokenizer(br.readLine(), " ");
    nv = Integer.parseInt(st.nextToken());
    int np = Integer.parseInt(st.nextToken());
    int v1, v2, cost;
    for (int i = 0; i < nv+1; ++i) {
      edges.add(new ArrayList<Node>());
    }

    for (int i = 0; i < nv-1; ++i) {
      st = new StringTokenizer(br.readLine(), " ");
      v1 = Integer.parseInt(st.nextToken());
      v2 = Integer.parseInt(st.nextToken());
      cost = Integer.parseInt(st.nextToken());
      edges.get(v1).add(new Node(v2, cost));
      edges.get(v2).add(new Node(v1, cost));
    }

    for (int i = 0; i < np; ++i) {
      st = new StringTokenizer(br.readLine(), " ");
      v1 = Integer.parseInt(st.nextToken());
      v2 = Integer.parseInt(st.nextToken());
      bw.write(bfs(v1, v2) + "\n");
    }
    bw.flush();
    bw.close();
    br.close();
  }
}
